# PCA Feature Transformation
Loads `X_features.npy`, applies full PCA (all components), and saves the result to `XPCA_features.npy`.

In [1]:
import numpy as np
from sklearn.decomposition import PCA
import time

## 1. Load Data

In [2]:
X = np.load("X_features.npy")
y = np.load("y_labels.npy")

print(f"X shape : {X.shape}")
print(f"y shape : {y.shape}")

X shape : (9689, 8110)
y shape : (9689,)


## 2. Fit PCA (all components)

**Why `n_components = min(n_samples, n_features)`?**  
PCA can produce at most `min(4844, 8110) = 4844` meaningful components — one per sample.  
Setting `n_components=None` would default to the same value but being explicit is cleaner.

In [3]:
n_components = min(X.shape[0], X.shape[1])   # 4844
print(f"Fitting PCA with n_components = {n_components} ...")

t0 = time.time()
pca = PCA(n_components=n_components, svd_solver="full", random_state=42)
X_pca = pca.fit_transform(X)
elapsed = time.time() - t0

print(f"Done in {elapsed:.1f}s")
print(f"X_pca shape : {X_pca.shape}")

Fitting PCA with n_components = 8110 ...
Done in 319.4s
X_pca shape : (9689, 8110)


## 3. Explained Variance Summary

In [4]:
cumvar = np.cumsum(pca.explained_variance_ratio_)

for threshold in [0.80, 0.90, 0.95, 0.99]:
    n = np.searchsorted(cumvar, threshold) + 1
    print(f"{int(threshold*100)}% variance explained by first {n:>5} components")

print(f"\nTotal components saved : {X_pca.shape[1]}")

80% variance explained by first   719 components
90% variance explained by first  1247 components
95% variance explained by first  1742 components
99% variance explained by first  2968 components

Total components saved : 8110


## 4. Save Results

In [5]:
np.save("XPCA_features.npy", X_pca)
print("Saved → XPCA_features.npy")

# Quick sanity check
loaded = np.load("XPCA_features.npy")
print(f"Verified shape : {loaded.shape}")

Saved → XPCA_features.npy
Verified shape : (9689, 8110)
